APCAppCore/APCAppCore/Library/Scoring/APCScoring.m

```diff
@@ -575,113 +735,148 @@ - (NSArray *)retrieveResultSummaryFromResults:(NSSet *)results latestOnly:(BOOL)
     }
     
-    // resort the grouped dataset by date
-    NSSortDescriptor *sortByDate = [[NSSortDescriptor alloc] initWithKey:kDatasetDateKey
-                                                               ascending:YES];
-    [groupedDataset sortUsingDescriptors:@[sortByDate]];
-    
-    [self.dataPoints removeAllObjects];
-    [self.dataPoints addObjectsFromArray:groupedDataset];
+    return summarizedDataset;
 }
 
-- (void)groupDatasetbyPeriod:(APHTimelineGroups)period
+- (NSDictionary *)groupByKeyPath:(NSString *)key dataset:(NSArray *)dataset
 {
-    NSDateComponents *groupDateComponents = [[NSDateComponents alloc] init];
+    NSMutableDictionary *groups = [NSMutableDictionary new];
     
-    for (NSDate *groupStartDate in self.timeline) {
-        NSDate *groupEndDate   = nil;
+    for (id object in dataset) {
+        id value = [object valueForKeyPath:key];
+        NSInteger yearNumber = [object[kDatasetGroupByYear] integerValue];
+        NSDateComponents *components = nil;
         
-        // Set start and end date for the grouping period
-        NSInteger weekNumber  = [[[NSCalendar currentCalendar] components:NSCalendarUnitWeekOfYear fromDate:groupStartDate] weekOfYear];
-        NSInteger monthNumber = [[[NSCalendar currentCalendar] components:NSCalendarUnitMonth fromDate:groupStartDate] month];
-        NSInteger yearNumber  = [[[NSCalendar currentCalendar] components:NSCalendarUnitYear fromDate:groupStartDate] year];
-        
-        if (period == APHTimelineGroupWeek) {
-            groupDateComponents.weekday = 7;
-            groupDateComponents.weekOfYear = weekNumber;
-            groupDateComponents.year = yearNumber;
-            
-            groupEndDate = [[NSCalendar currentCalendar] dateFromComponents:groupDateComponents];
-        } else if (period == APHTimelineGroupMonth) {
-            groupDateComponents.month = monthNumber;
-            groupEndDate = [[NSCalendar currentCalendar] dateFromComponents:groupDateComponents];
-        } else { // defaults to Day
-            groupEndDate = groupStartDate;
+        if ([key isEqualToString:kDatasetGroupByMonth]) {
+            components = [[NSDateComponents alloc] init];
+            components.weekday = 1;
+            components.month = [value integerValue];
+            components.year = yearNumber;
+        } else if ([key isEqualToString:kDatasetGroupByYear]) {
+            components = [[NSDateComponents alloc] init];
+            components.weekday = 1;
+            components.year = [value integerValue];
+        } else if ([key isEqualToString:kDatasetGroupByWeek]) {
+            components = [[NSDateComponents alloc] init];
+            components.weekday = 1;
+            components.weekOfYear = [value integerValue];
+            components.year = yearNumber;
+        } else {
+            // Group by day
+            components = [[NSCalendar currentCalendar] components:(NSCalendarUnitDay|NSCalendarUnitMonth|NSCalendarUnitYear)
```

Comment:
I guess perhaps we don't care about localization yet, but this won't be sufficient in calendars that use era (e.g., Japanese)